ocaml/typing/rec_check.ml

```diff
@@ -1192,13 +1194,16 @@ and case
        ----------------------------------------
        G - p; m[mp] |- (p (when g)? -> e) : m
     *)
-    let judg = join [
-        option expression c_guard << Dereference;
-        expression c_rhs;
-      ] in
-    (fun m ->
-       let env = judg m in
-       (remove_pat c_lhs env), Mode.compose m (pattern c_lhs env))
+      let judg = match c_guard with
+        | None -> expression c_rhs
+        | Some (Predicate p) ->
+            join [ expression p << Dereference; expression c_rhs ]
+        | Some (Pattern (e, _, pat)) ->
+          let cases = [ { c_lhs = pat; c_guard = None; c_rhs } ] in
+          check_match e cases in
+      (fun m ->
```

Comment:
Let's:
  * add a comment on `check_match` relating it back tot he `expression` judgment
  * probably also rename it to `match_`? (this matches what we do elsewhere.) or `match_expression` might be better
  * comment on `match_expression` saying that it's used for pattern guards too
  * triplicate the comment for the guard cases and fix up the rule in each one

Reply:
Done! Thanks for the help in figuring out the judgement stuff